### PR TITLE
 More precise includes of `Node` constants

### PR DESCRIPTION
**Pull request by 4tXJ7f**
We store constants, e.g., `BitVector` and `Rational`, in our node infrastructure. As a result, we were indirectly including some headers in almost all files, e.g., the GMP headers. This commit changes that by forward-declaring the classes for the constants. As a result, we have to include headers like `util/rational.h` explicitly when we use `Rational` but it saves about 3 minutes in compile time (CPU time).

The commit changes `RoundingMode` from an `enum` to an `enum class` such that it can be forward declared.

### CODE REVIEW

File: src/preprocessing/passes/bv_to_int.h [hunk 1]
```diff
@@ -73,6 +73,7 @@
 #include "context/cdhashset.h"
 #include "preprocessing/preprocessing_pass.h"
 #include "theory/arith/nl/iand_utils.h"
+#include "util/rational.h"
```
Comment (mpreiner) [inline src/preprocessing/passes/bv_to_int.h:76]: Rational doesn't seem to be used in this header, include shouldn't be needed here.

File: src/theory/bv/kinds [hunk 2]
```diff
@@ -13,7 +13,8 @@ properties check propagate presolve ppStaticLearn
 rewriter ::cvc5::theory::bv::TheoryBVRewriter "theory/bv/theory_bv_rewriter.h"
 
 constant BITVECTOR_TYPE \
-	::cvc5::BitVectorSize \
+  struct \
+	BitVectorSize \
```
Comment (nafur) [inline src/theory/bv/kinds:17]:
```suggestion
      struct \
	BitVectorSize \
```

File: src/theory/bv/kinds [hunk 3]
```diff
@@ -101,56 +103,64 @@ operator BITVECTOR_EAGER_ATOM 1 "formula to be treated as a bv atom via eager bi
 ### parameterized operator kinds ----------------------------------------------
 
 constant BITVECTOR_BITOF_OP \
-	::cvc5::BitVectorBitOf \
+  struct \
+	BitVectorBitOf \
```
Comment (nafur) [inline src/theory/bv/kinds:107]:
```suggestion
      struct \
	BitVectorBitOf \
```

Comment (nafur) [reply src/theory/bv/kinds:107]: and more below...

**Review by nafur — CHANGES_REQUESTED** [review]
A few minor comments.

File: src/expr/type_node.cpp [hunk 4]
```diff
@@ -684,6 +687,18 @@ TypeNode TypeNode::getBagElementType() const
...
+  return (getKind() == kind::BITVECTOR_TYPE
+          && getConst<BitVectorSize>() == size);
+}
+
+unsigned TypeNode::getBitVectorSize() const
+{
+  Assert(isBitVector());
+  return getConst<BitVectorSize>();
+}
+
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/expr/type_node.cpp:701]: I agree, it would be good to change this in this PR.

**Review by mpreiner — APPROVED** [review]

**Comment by 4tXJ7f** [comment]
Thanks for the comments. I have addressed them and updated my vim config to show me tabs as characters ;)

**Review by nafur — APPROVED** [review]
LGTM.

**Pull request merged**